pkg/controller/registrysyncer/registrysyncer.go

```diff
@@ -0,0 +1,336 @@
+	for clusterName, isTag := range isTags {
+		if t.Before(&isTag.Image.CreationTimestamp) {
+			t = isTag.Image.CreationTimestamp
+			resultContext = clusterName
+		}
+	}
+	return resultContext
+}
+
+const pullSecretName = "registry-cluster-pull-secret"
+
+func (r *reconciler) ensureImagePullSecret(namespace string, client ctrlruntimeclient.Client, log *logrus.Entry) error {
+	secret, mutateFn := r.pullSecret(namespace)
+	return upsertObject(r.ctx, client, secret, mutateFn, log)
+}
+
+// https://issues.redhat.com/browse/DPTP-1656?focusedCommentId=15345756&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-15345756
+// E.g., ci-operator uses the release controller configuration to determine
+// the version of OpenShift we create from the ImageStream, so we need
+// to copy the annotation if it exists
+const releaseConfigAnnotationPrefix = "release.openshift.io"
+
+func imagestream(imageStream *imagev1.ImageStream) (*imagev1.ImageStream, crcontrollerutil.MutateFn) {
+	stream := &imagev1.ImageStream{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: imageStream.Namespace,
+			Name:      imageStream.Name,
+		},
+	}
+	return stream, func() error {
+		for k, v := range imageStream.Annotations {
+			if stream.Annotations == nil {
+				stream.Annotations = map[string]string{}
+			}
+			if strings.HasPrefix(k, releaseConfigAnnotationPrefix) {
+				stream.Annotations[k] = v
+			}
+		}
+		stream.Spec.LookupPolicy.Local = true
+		for i := range stream.Spec.Tags {
+			stream.Spec.Tags[i].ReferencePolicy.Type = imagev1.LocalTagReferencePolicy
+		}
+		return nil
+	}
+}
+
+func (r *reconciler) ensureImageStream(imageStream *imagev1.ImageStream, client ctrlruntimeclient.Client, log *logrus.Entry) error {
+	stream, mutateFn := imagestream(imageStream)
+	return upsertObject(r.ctx, client, stream, mutateFn, log)
+}
+
+func (r *reconciler) pullSecret(namespace string) (*corev1.Secret, crcontrollerutil.MutateFn) {
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      pullSecretName,
+		},
+	}
+	return s, func() error {
+		s.Data = map[string][]byte{
+			corev1.DockerConfigJsonKey: r.pullSecretGetter(),
+		}
+		s.Type = corev1.SecretTypeDockerConfigJson
+		return nil
+	}
+}
+
+func testInputImageStreamTagFilterFactory() objectFilter {
+	return func(nn types.NamespacedName) bool {
+		// TODO: make the filter for deny-list
+		return true
+	}
+}
+
+func publicDomainForImage(clusterName, potentiallyPrivate string) (string, error) {
+	d, err := domainForContextAndService(clusterName, api.ServiceRegistry)
+	if err != nil {
+		return "", err
+	}
+	svcDomainAndPort := "image-registry.openshift-image-registry.svc:5000"
+	if clusterName == "api.ci" {
+		svcDomainAndPort = "docker-registry.default.svc:5000"
+	}
+
+	return strings.ReplaceAll(potentiallyPrivate, svcDomainAndPort, d), nil
+}
+
+func domainForContextAndService(context string, service api.Service) (string, error) {
```

Comment:
The second arg to this func is useless, it can only be `api.ServiceRegistry` which then means you can get rid of the inner switch statement below

Reply:
Done.